main/src/library/Core/Io/IOError.flix

```diff
@@ -0,0 +1,14 @@
+namespace Core/Io/IOError {
+    ///
+    /// Represents an error when operating in an I/O context.
+    ///
+    pub enum IOError {
+        // Generalized
+        case ElementNotFound
+        case FileNotFound
+        case Generic
+        case IllegalState
+        case IncorrectBufferSize
+        case MalformedZip
+    }
```

Comment:
We need a description of each error.

Reply:
Each error should carry a string with more information. Otherwise it will be possible to debug the error.

Reply:
> Why are we not using the error message from the exception? Is it because its broken?

When I started writing these wrappers the only IO-related error handling was in `Path.flix`, and that amounted to returning the same error string no matter the error that happened and only catching `IOException`s. Since the implementor of `Path.flix` hadn't used them I assumed there was a reason - design-wise or opinion-wise - for why they weren't used.

> The more I think about this, the more unsatisfied I am with using an enum.

I'd be interested to hear exactly what makes you unsatisfied about doing error handling via an enum. Bringing up Rust, it's one of the things the language has gotten a lot of praise for. Further, to me it seems like the logical extension of using ADTs to represent computations that can fail - i.e. `Result`.

> Like what about just a string or just a record?

I think going to a string for representing errors would be a very wrong move. The experience of having to consume errors that are solely string-based would be roughly equivalent to every Java program operating with only plain `Exception` objects.

With a record I see similar issues. You would get, I assume, more granular information that having it all in a String, but you're still not getting any information on what kind of error occurred - at least not in a way that is accessible ergonomically. 

> Note that in Rust `IOError` is separated from `ErrorKind`. Not sure what to make of that design.

Specifically the `io::Error` enum from Rust is a bit different than what normal errors look like in the language. I'm not sure if you looked at the implementation. `io::Error` is an enum that can either be an integer error code from the os, a "simple" error that only contains an `ErrorKind` (this is what I've implemented currently), and finally there is also a "custom" error variant where it contains an `ErrorKind` and another implementor of the `Error` trait. The last variant essentially has an `err.inner`, which can be something as simple as a String. 

The last variant of the `io::Error` type is not possible to implement in Flix while keeping `Result` composable - we again run into the problem of [painting functions](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/). In Rust it's possible to box "trait objects" which results in a vtable-style execution. That wouldn't be possible in Flix without doing a kludge to subvert the type system entirely.